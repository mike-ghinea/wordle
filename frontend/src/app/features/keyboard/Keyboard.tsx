import s from "./Keyboard.styles";
import { addLetter, removeLetter } from "../word/wordSlice";
import { useSelector } from "react-redux";
import { RootState } from "../../store";
import { useAppDispatch } from "../../hooks";
import { handleEnter } from "../common/utils";

const getUppercase = (key: string) => {
  switch (key) {
    case "↵":
      return "ENTER";
    case "←":
      return "DELETE";
    default:
      return key.toUpperCase();
  }
};

const getAriaLabel = (key: string) => {
  switch (key) {
    case "↵":
      return "enter";
    case "←":
      return "backspace";
    default:
      return "add " + key;
  }
};

const getFunction = (
  word: string,
  key: string,
  dispatch: ReturnType<typeof useAppDispatch>,
) => {
  switch (key) {
    case "↵":
      return () => handleEnter(word, dispatch);
    case "←":
      return () => {
        if (word.length === 0) return;
        dispatch(removeLetter());
      };
    default:
      return () => {
        if (word.length > 4) return;
        dispatch(addLetter(key));
      };
  }
};

const KeyboardRow: React.FC<{ letters: string }> = ({ letters }) => {
  const word = useSelector((state: RootState) => state.word.grid);
  const row = useSelector((state: RootState) => state.word.row);
  const answerGrid = useSelector((state: RootState) => state.word.answerGrid);
  const disabledLetters = useSelector(
    (state: RootState) => state.word.disabledLetters,
  );
  const correctLetters = useSelector(
    (state: RootState) => state.word.correctLetters,
  );
  const wrongLocationLetters = useSelector(
    (state: RootState) => state.word.wrongLocationLetters,
  );
  const dispatch = useAppDispatch();

  const disableKeyboard = row > 5 || answerGrid[row - 1]?.join("") === "ggggg";

  return (
    <s.KeyboardRowWrapper>
      {letters.split("").map((letter) => (
        <s.KeyboardCell
          key={letter}
          data-key={letter}
          aria-label={getAriaLabel(letter)}
          onClick={
            disableKeyboard
              ? undefined
              : getFunction(word[row], letter, dispatch)
          }
          $disabled={disabledLetters.includes(letter)}
          $correct={correctLetters.includes(letter)}
          $wrongLocation={wrongLocationLetters.includes(letter)}
        >
          {getUppercase(letter)}
        </s.KeyboardCell>
      ))}
    </s.KeyboardRowWrapper>
  );
};

export const Keyboard = () => {
  return (
    <s.KeyboardWrapper>
      <KeyboardRow letters="qwertyuiop" />
      <KeyboardRow letters="asdfghjkl" />
      <KeyboardRow letters="↵zxcvbnm←" />
    </s.KeyboardWrapper>
  );
};
