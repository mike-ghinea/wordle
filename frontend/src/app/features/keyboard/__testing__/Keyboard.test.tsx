import { Keyboard } from "../Keyboard";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import wordReducer from "../../../features/word/wordSlice";
import { act, render } from "@testing-library/react";

const createTestStore = () =>
  configureStore({
    reducer: {
      word: wordReducer,
    },
  });

describe("Keyboard", () => {
  let store: ReturnType<typeof createTestStore>;

  beforeEach(() => {
    store = createTestStore();
  });

  it("renders Keyboard correctly", () => {
    const { baseElement } = render(
      <Provider store={store}>
        <Keyboard />
      </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it("should handle keypress correctly - letter", () => {
    const { getByText } = render(
      <Provider store={store}>
        <Keyboard />
      </Provider>,
    );
    const key = getByText("A");
    act(() => {
      key.click();
    });
    expect(store.getState().word.grid[0]).toBe("a");
  });

  it("should handle keypress correctly - backspace", () => {
    const { getByText } = render(
      <Provider store={store}>
        <Keyboard />
      </Provider>,
    );
    let key = getByText("A");
    act(() => {
      key.click();
    });
    expect(store.getState().word.grid[0]).toBe("a");

    key = getByText("DELETE");
    act(() => {
      key.click();
    });
    expect(store.getState().word.grid[0]).toBe("");
  });

  it("should handle keypress correctly - enter", () => {
    const { getByText } = render(
      <Provider store={store}>
        <Keyboard />
      </Provider>,
    );
    let key = getByText("A");
    act(() => {
      key.click();
    });
    expect(store.getState().word.grid[0]).toBe("a");

    key = getByText("ENTER");
    act(() => {
      key.click();
    });
    expect(store.getState().word.grid[0]).toBe("a");
  });
});
