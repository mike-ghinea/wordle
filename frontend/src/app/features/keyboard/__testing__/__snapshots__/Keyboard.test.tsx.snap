// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders Keyboard correctly 1`] = `
<body>
  <div>
    <div
      class="sc-FEMpB jlUFLC"
    >
      <div
        class="sc-gjcoXW jZoyXw"
      >
        <button
          aria-label="add q"
          class="sc-dVBluf gTajsM"
          data-key="q"
        >
          Q
        </button>
        <button
          aria-label="add w"
          class="sc-dVBluf gTajsM"
          data-key="w"
        >
          W
        </button>
        <button
          aria-label="add e"
          class="sc-dVBluf gTajsM"
          data-key="e"
        >
          E
        </button>
        <button
          aria-label="add r"
          class="sc-dVBluf gTajsM"
          data-key="r"
        >
          R
        </button>
        <button
          aria-label="add t"
          class="sc-dVBluf gTajsM"
          data-key="t"
        >
          T
        </button>
        <button
          aria-label="add y"
          class="sc-dVBluf gTajsM"
          data-key="y"
        >
          Y
        </button>
        <button
          aria-label="add u"
          class="sc-dVBluf gTajsM"
          data-key="u"
        >
          U
        </button>
        <button
          aria-label="add i"
          class="sc-dVBluf gTajsM"
          data-key="i"
        >
          I
        </button>
        <button
          aria-label="add o"
          class="sc-dVBluf gTajsM"
          data-key="o"
        >
          O
        </button>
        <button
          aria-label="add p"
          class="sc-dVBluf gTajsM"
          data-key="p"
        >
          P
        </button>
      </div>
      <div
        class="sc-gjcoXW jZoyXw"
      >
        <button
          aria-label="add a"
          class="sc-dVBluf gTajsM"
          data-key="a"
        >
          A
        </button>
        <button
          aria-label="add s"
          class="sc-dVBluf gTajsM"
          data-key="s"
        >
          S
        </button>
        <button
          aria-label="add d"
          class="sc-dVBluf gTajsM"
          data-key="d"
        >
          D
        </button>
        <button
          aria-label="add f"
          class="sc-dVBluf gTajsM"
          data-key="f"
        >
          F
        </button>
        <button
          aria-label="add g"
          class="sc-dVBluf gTajsM"
          data-key="g"
        >
          G
        </button>
        <button
          aria-label="add h"
          class="sc-dVBluf gTajsM"
          data-key="h"
        >
          H
        </button>
        <button
          aria-label="add j"
          class="sc-dVBluf gTajsM"
          data-key="j"
        >
          J
        </button>
        <button
          aria-label="add k"
          class="sc-dVBluf gTajsM"
          data-key="k"
        >
          K
        </button>
        <button
          aria-label="add l"
          class="sc-dVBluf gTajsM"
          data-key="l"
        >
          L
        </button>
      </div>
      <div
        class="sc-gjcoXW jZoyXw"
      >
        <button
          aria-label="enter"
          class="sc-dVBluf gTajsM"
          data-key="↵"
        >
          ENTER
        </button>
        <button
          aria-label="add z"
          class="sc-dVBluf gTajsM"
          data-key="z"
        >
          Z
        </button>
        <button
          aria-label="add x"
          class="sc-dVBluf gTajsM"
          data-key="x"
        >
          X
        </button>
        <button
          aria-label="add c"
          class="sc-dVBluf gTajsM"
          data-key="c"
        >
          C
        </button>
        <button
          aria-label="add v"
          class="sc-dVBluf gTajsM"
          data-key="v"
        >
          V
        </button>
        <button
          aria-label="add b"
          class="sc-dVBluf gTajsM"
          data-key="b"
        >
          B
        </button>
        <button
          aria-label="add n"
          class="sc-dVBluf gTajsM"
          data-key="n"
        >
          N
        </button>
        <button
          aria-label="add m"
          class="sc-dVBluf gTajsM"
          data-key="m"
        >
          M
        </button>
        <button
          aria-label="backspace"
          class="sc-dVBluf gTajsM"
          data-key="←"
        >
          DELETE
        </button>
      </div>
    </div>
  </div>
</body>
`;
